### Importing libraries

In [22]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
from string import ascii_lowercase
from numpy.linalg import inv
from math import sqrt
%matplotlib inline

### loading Data

In [4]:
#reading data files
store_df=pd.read_csv("store.csv")
train_df=pd.read_csv("train.csv")

C:\Users\fahad\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3072: DtypeWarning: Columns (7) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### marging 2 dataframes which will help us later

In [6]:
#Merging both the Dataframes into one based on the "Store" ID
df=store_df.merge(train_df,on=["Store"],how="inner")


### sorting data in ascending order  based on date

In [8]:
df2 = df.sort_values('Date',ascending=True)

### aranging data such that I have store on x-asix and days on y-axis

In [10]:
c=1
dict1 = {}
for i in range(1,1116):
    dict1[c] = df2[df2["Store"]==i]["Sales"].values
    c=c+1
    
df_3 = pd.DataFrame.from_dict(dict1,orient='index')
df_4 = df_3.T
df_4.index = df_4.index + 1
df_5 = df_4.T
df_5

1        2        3        4        5    6        7        8        9    \
1     0.0   5530.0   4327.0   4486.0   4997.0  0.0   7176.0   5580.0   5471.0   
2     0.0   4422.0   4159.0   4484.0   2342.0  0.0   6775.0   6318.0   6763.0   
3     0.0   6823.0   5902.0   6069.0   4523.0  0.0  12247.0   9800.0   8001.0   
4     0.0   9941.0   8247.0   8290.0  10338.0  0.0  12112.0  10031.0   8857.0   
5     0.0   4253.0   3465.0   4456.0   1590.0  0.0   6978.0   5718.0   5974.0   
...   ...      ...      ...      ...      ...  ...      ...      ...      ...   
1111  0.0   5097.0   4579.0   4640.0   3325.0  0.0   9444.0   6472.0   5307.0   
1112  0.0  10797.0   8716.0   9788.0   9513.0  0.0  25165.0  17058.0  14724.0   
1113  0.0   6218.0   5563.0   5524.0   5194.0  0.0   8984.0   6866.0   6115.0   
1114  0.0  20642.0  18463.0  18371.0  18856.0  0.0  21237.0  18816.0  17073.0   
1115  0.0   3697.0   4297.0   4540.0   4771.0  0.0   6905.0   5243.0   4649.0   

          10   ...      933      934      935      936  937      938      939  \
1      4892.0  ...   3464.0   3769.0   3706.0   4364.0  0.0   6102.0   5011.0   
2      5618.0  ...   5093.0   4108.0   3854.0   2512.0  0.0   6627.0   5671.0   
3      7772.0  ...   5414.0   5702.0   5080.0   3878.0  0.0   8107.0   8864.0   
4      9472.0  ...   8503.0   7286.0   8322.0   9322.0  0.0  11812.0  10275.0   
5      4999.0  ...   3595.0   3713.0   3815.0   2030.0  0.0   7059.0   6083.0   
...       ...  ...      ...      ...      ...      ...  ...      ...      ...   
1111   5887.0  ...   4021.0   3587.0   3918.0   2177.0  0.0   7742.0   6793.0   
1112  14366.0  ...   6029.0   6730.0   6220.0   6216.0  0.0  14383.0   9583.0   
1113   7508.0  ...   4565.0   6410.0   6399.0   4784.0  0.0   7582.0   6468.0   
1114  18075.0  ...  20424.0  20564.0  19627.0  21312.0  0.0  26720.0  25518.0   
1115   5007.0  ...   5342.0   6150.0   5816.0   6897.0  0.0  10712.0   8093.0   

          940      941      942  
1      4782.0   5020.0   5263.0  
2      6402.0   5567.0   6064.0  
3      7610.0   8977.0   8314.0  
4     10514.0  10387.0  13995.0  
5      5899.0   4943.0   4822.0  
...       ...      ...      ...  
1111   4907.0   5263.0   5723.0  
1112   9179.0   9652.0   9626.0  
1113   6640.0   7491.0   7289.0  
1114  25840.0  24395.0  27508.0  
1115   7661.0   8405.0   8680.0  

[1115 rows x 942 columns]

In [12]:
df_5.shape

(1115, 942)

## B,C .)Test train split as described in exercise, size of sets

In [13]:
X_train = df_5.iloc[:1000,:900]
Y_train = df_5.iloc[:1000,900:]

X_test = df_5.iloc[1000:,:900]
Y_test = df_5.iloc[1000:,900:]

### Fillling nan values with 0

In [14]:
X_train.fillna(0,inplace=True)
Y_train.fillna(0,inplace=True)
X_test.fillna(0,inplace=True)
Y_test.fillna(0,inplace=True)


## D.) learning beta_hat on X_train using Y_train

In [15]:
X_transpose = X_train.T
A = np.dot(X_transpose , X_train)

# beta_All = []
beta_All = np.zeros(shape=(42,900))
for i in range(42):
    b = np.dot(X_transpose , Y_train.iloc[:,i])
    inverse_A = inv(A)
    beta_hat = np.dot(inverse_A,b)
    beta_All[i] = beta_hat


In [16]:
beta_All.shape

(42, 900)

### Doing prediction using X_train and beta_all

In [18]:
predictions = np.dot(X_train,beta_All.T)

In [19]:
predictions.shape

(1000, 42)

### Saving errors on X_train

In [20]:
erros = np.zeros(shape=(42,2))

In [23]:
for i in range(42):
    numerator = ((Y_train.iloc[:,i] - predictions[:,i])**2).sum()
    rmse = sqrt((numerator)/predictions.shape[0])
    erros[i][0] = rmse
    
    numerator_mae = (np.absolute(Y_train.iloc[:,i] - predictions[:,i])).sum()
    mae = numerator_mae/predictions.shape[0]
    erros[i][1] = mae
error_df = pd.DataFrame({'RMSE': erros[:, 0], 'MAE': erros[:, 1]})
error_df.index = error_df.index + 1

In [24]:
error_df ## error of next 42 days

RMSE           MAE
1   17742.178410   7261.079762
2   25051.453698  10488.065678
3   23383.945729   9078.430441
4   16374.261714   7579.979019
5   15792.176309   7337.978031
6   35813.619679  10039.490169
7   16946.616113   7934.038644
8   20188.112708   7651.088404
9   45364.399622  13189.359547
10  29546.166599  11377.884366
11  48787.265779  13834.538454
12  21704.428328   9832.822747
13  39610.263478  12176.949307
14  23594.244903  10256.222223
15  20594.617055   7814.542089
16  28283.166187  11223.705382
17  20273.061667   8970.041429
18  16693.092280   7790.156637
19  23649.669260   9197.559170
20  20092.818815   8479.333529
21  27658.760352   9765.938855
22  27335.766120   8576.323706
23  22550.255314  10486.482724
24  21658.215083  10082.169299
25  20658.380343   9553.922841
26  21072.126226   9521.620328
27  19118.083562   9023.237637
28  21735.012922   9606.134782
29  17276.286004   7571.165500
30  39950.592120  13320.322665
31  18345.540910   8396.796350
32  22148.422581   8506.892745
33  29555.217153   9326.345512
34  26717.834783   9462.702496
35  27800.650653   9435.990830
36  18020.188318   7204.522519
37  34954.152934  12543.830844
38  28416.186118  10946.520104
39  25946.001798  10103.812515
40  30669.862917  10513.181563
41  19196.532250   9041.647051
42  31681.791291  11664.842948

In [28]:
print(f'RMSE of all next 42 days: {error_df["RMSE"].sum()} on X_train')
print(f'MAE of all next 42 days: {error_df["MAE"].sum()} on X_train')

RMSE of all next 42 days: 1061951.4180856112 on X_train
RMSE of all next 42 days: 406167.66884399526 on X_train


### Doing prediction on X_test using beta_all learned previously. 

In [40]:
predictions_Test = np.dot(X_test,beta_All.T)

### Saving errors on X_test

In [41]:
error_test_set= np.zeros(shape=(42,2))

In [42]:
for i in range(42):
    numerator = ((Y_test.iloc[:,i] - predictions_Test[:,i])**2).sum()
    rmse = sqrt((numerator)/predictions_Test.shape[0])
    error_test_set[i][0] = rmse
    
    numerator_mae = (np.absolute(Y_test.iloc[:,i] - predictions_Test[:,i])).sum()
    mae = numerator_mae/predictions_Test.shape[0]
    error_test_set[i][1] = mae
error_test_df = pd.DataFrame({'RMSE': error_test_set[:, 0], 'MAE': error_test_set[:, 1]})
error_test_df.index = error_test_df.index+1

In [43]:
error_test_df ## error of next 42 days using X_Test

RMSE            MAE
1   1.910052e+06  285644.220354
2   2.486968e+06  341309.090865
3   2.339812e+06  345599.340580
4   2.183981e+06  323685.078308
5   2.174691e+06  322919.538776
6   2.167838e+06  323575.430430
7   2.338801e+06  346507.917570
8   2.010460e+06  302062.422584
9   2.577042e+06  355210.156034
10  3.172999e+06  483580.318826
11  3.206733e+06  486822.572481
12  2.940033e+06  439199.652093
13  2.864946e+06  428159.266454
14  2.877099e+06  428524.709966
15  2.028034e+06  301386.111873
16  2.617610e+06  359445.451311
17  2.556167e+06  380457.888680
18  2.287479e+06  339190.967384
19  2.369299e+06  351420.535493
20  2.356235e+06  351167.001326
21  2.424344e+06  359122.390078
22  2.033275e+06  305222.505990
23  2.684876e+06  367807.230785
24  3.044006e+06  459417.334152
25  2.842539e+06  426903.478132
26  2.744141e+06  410457.487664
27  2.702535e+06  403814.268279
28  2.693902e+06  399166.136231
29  2.125660e+06  317725.616122
30  2.785025e+06  379617.142660
31  2.466937e+06  367615.413598
32  2.263567e+06  336930.133952
33  2.240019e+06  333297.254603
34  2.286355e+06  338589.901579
35  2.339103e+06  347570.626973
36  1.931075e+06  287910.180313
37  2.697665e+06  367210.613405
38  3.002125e+06  454242.340523
39  2.751553e+06  413665.230208
40  2.651461e+06  398821.811981
41  2.723998e+06  407532.475885
42  3.085981e+06  463390.400036

In [44]:
print(f'RMSE of all next 42 days: {error_test_df["RMSE"].sum()} on X_test')
print(f'MAE of all next 42 days: {error_test_df["MAE"].sum()} on X_test')

RMSE of all next 42 days: 105986420.26405494 on X_test
MAE of all next 42 days: 15641897.644539073 on X_test


<hr style="border:2px solid gray"> </hr>

### g.1)

### Getting the last recorded sales value of each store and then repeating it for the next 42 days to get Y_train.

In [46]:
Y_train_last_recorded_sale = pd.DataFrame(X_train.iloc[:,-1])

# where n is the number of times  we want to duplicate the column
Y_train_last__sale = Y_train_last_recorded_sale.assign(**{str(x): Y_train_last_recorded_sale.iloc[:, 0] for x in range(2,43)})
Y_train_last__sale.rename(columns={900:1}, inplace=True )
Y_train_last__sale.shape

(1000, 42)

### learning beta_all on X_train using new Y_train

In [47]:
X_transpose = X_train.T
A = np.dot(X_transpose , X_train)

beta_All = np.zeros(shape=(42,900))
for i in range(42):
    b = np.dot(X_transpose , Y_train_last__sale.iloc[:,i])
    inverse_A = inv(A)
    beta_hat = np.dot(inverse_A,b)
    beta_All[i] = beta_hat


### Doing prediction using X_train and beta_all, Then finding both rmse and mae error and display it

In [49]:
predictions = np.dot(X_train,beta_All.T)

erros = np.zeros(shape=(42,2))
for i in range(42):
    numerator = ((Y_train_last__sale.iloc[:,i] - predictions[:,i])**2).sum()
    rmse = sqrt((numerator)/predictions.shape[0])
    erros[i][0] = rmse
    
    numerator_mae = (np.absolute(Y_train_last__sale.iloc[:,i] - predictions[:,i])).sum()
    mae = numerator_mae/predictions.shape[0]
    
    erros[i][1] = mae
    
error_df = pd.DataFrame({'RMSE': erros[:, 0], 'MAE': erros[:, 1]})
error_df.index = error_df.index + 1
error_df

RMSE           MAE
1   36342.178753  11433.101447
2   36342.178753  11433.101447
3   36342.178753  11433.101447
4   36342.178753  11433.101447
5   36342.178753  11433.101447
6   36342.178753  11433.101447
7   36342.178753  11433.101447
8   36342.178753  11433.101447
9   36342.178753  11433.101447
10  36342.178753  11433.101447
11  36342.178753  11433.101447
12  36342.178753  11433.101447
13  36342.178753  11433.101447
14  36342.178753  11433.101447
15  36342.178753  11433.101447
16  36342.178753  11433.101447
17  36342.178753  11433.101447
18  36342.178753  11433.101447
19  36342.178753  11433.101447
20  36342.178753  11433.101447
21  36342.178753  11433.101447
22  36342.178753  11433.101447
23  36342.178753  11433.101447
24  36342.178753  11433.101447
25  36342.178753  11433.101447
26  36342.178753  11433.101447
27  36342.178753  11433.101447
28  36342.178753  11433.101447
29  36342.178753  11433.101447
30  36342.178753  11433.101447
31  36342.178753  11433.101447
32  36342.178753  11433.101447
33  36342.178753  11433.101447
34  36342.178753  11433.101447
35  36342.178753  11433.101447
36  36342.178753  11433.101447
37  36342.178753  11433.101447
38  36342.178753  11433.101447
39  36342.178753  11433.101447
40  36342.178753  11433.101447
41  36342.178753  11433.101447
42  36342.178753  11433.101447

In [51]:
print(f'RMSE of all next 42 days: {error_df["RMSE"].sum()} on X_train')
print(f'MAE of all next 42 days: {error_df["MAE"].sum()} on X_train')

RMSE of all next 42 days: 1526371.507643836 on X_train
MAE of all next 42 days: 480190.2607947011 on X_train


### Doing prediction on X_test using beta_all learned previously. Then finding both rmse and mae error and display it

In [52]:
predictions_Test = np.dot(X_test,beta_All.T)
error_test_set= np.zeros(shape=(42,2))
for i in range(42):
    numerator = ((Y_test.iloc[:,i] - predictions_Test[:,i])**2).sum()
    rmse = sqrt((numerator)/predictions_Test.shape[0])
    error_test_set[i][0] = rmse
    
    numerator_mae = (np.absolute(Y_test.iloc[:,i] - predictions_Test[:,i])).sum()
    mae = numerator_mae/predictions_Test.shape[0]
    error_test_set[i][1] = mae
error_test_df = pd.DataFrame({'RMSE': error_test_set[:, 0], 'MAE': error_test_set[:, 1]})
error_test_df.index = error_test_df.index
error_test_df

RMSE            MAE
0   2.710415e+06  406273.012404
1   2.709998e+06  408726.793447
2   2.710386e+06  406190.601191
3   2.710350e+06  406215.173768
4   2.710382e+06  406233.049155
5   2.710350e+06  406268.712373
6   2.710338e+06  406131.639273
7   2.710452e+06  406260.493478
8   2.710102e+06  408732.549969
9   2.710664e+06  406061.890526
10  2.710541e+06  406051.563484
11  2.710451e+06  405888.367515
12  2.710541e+06  405897.070452
13  2.710425e+06  405884.025940
14  2.710446e+06  406319.116340
15  2.709946e+06  408714.749969
16  2.710700e+06  406072.239498
17  2.710479e+06  406164.362730
18  2.710557e+06  406054.937028
19  2.710509e+06  406086.020057
20  2.710414e+06  405943.988510
21  2.710537e+06  406364.507852
22  2.710029e+06  408718.741273
23  2.710699e+06  405914.419656
24  2.710551e+06  405805.920866
25  2.710546e+06  405947.386697
26  2.710626e+06  405979.864958
27  2.710598e+06  405905.445920
28  2.710544e+06  406265.464374
29  2.710206e+06  408732.889099
30  2.710652e+06  406078.124278
31  2.710610e+06  406265.324437
32  2.710379e+06  406219.715741
33  2.710748e+06  406214.948359
34  2.710482e+06  406098.611571
35  2.710550e+06  406461.690461
36  2.710182e+06  408729.923882
37  2.710876e+06  405945.499815
38  2.710617e+06  405840.772527
39  2.710651e+06  405866.235195
40  2.710723e+06  405904.910685
41  2.710656e+06  405775.975700

In [53]:
print(f'RMSE of all next 42 days: {error_test_df["RMSE"].sum()} on X_test')
print(f'MAE of all next 42 days: {error_test_df["MAE"].sum()} on X_test')

RMSE of all next 42 days: 113839908.54505162 on X_test
MAE of all next 42 days: 17071206.73044961 on X_test


<hr style="border:2px solid gray"> </hr><hr style="border:2px solid gray"> </hr>


### G.b

### Duplicating the mean of sales for the sales horizon to get Y_train.

In [54]:
Y_train_last_recorded_sale =  pd.DataFrame(X_train.mean(axis=1))

# where n is the number of columns you want
Y_train_last__sale = Y_train_last_recorded_sale.assign(**{str(x): Y_train_last_recorded_sale.iloc[:, 0] for x in range(2,43)})
Y_train_last__sale.rename(columns={0:1}, inplace=True )
Y_train_last__sale.shape

(1000, 42)

### learning beta_all on X_train using new Y_train

In [56]:
X_transpose = X_train.T
A = np.dot(X_transpose , X_train)

beta_All = np.zeros(shape=(42,900))
for i in range(42):
    b = np.dot(X_transpose , Y_train_last__sale.iloc[:,i])
    inverse_A = inv(A)
    beta_hat = np.dot(inverse_A,b)
    beta_All[i] = beta_hat



### Doing prediction using X_train and beta_all, Then finding both rmse and mae error and display it

In [57]:
predictions = np.dot(X_train,beta_All.T)

erros = np.zeros(shape=(42,2))
for i in range(42):
    numerator = ((Y_train_last__sale.iloc[:,i] - predictions[:,i])**2).sum()
    rmse = sqrt((numerator)/predictions.shape[0])
    erros[i][0] = rmse
    
    numerator_mae = (np.absolute(Y_train_last__sale.iloc[:,i] - predictions[:,i])).sum()
    mae = numerator_mae/predictions.shape[0]
    
    erros[i][1] = mae
    
error_df = pd.DataFrame({'RMSE': erros[:, 0], 'MAE': erros[:, 1]})
error_df.index = error_df.index + 1
error_df

RMSE          MAE
1   23711.973338  9173.530539
2   23711.973338  9173.530539
3   23711.973338  9173.530539
4   23711.973338  9173.530539
5   23711.973338  9173.530539
6   23711.973338  9173.530539
7   23711.973338  9173.530539
8   23711.973338  9173.530539
9   23711.973338  9173.530539
10  23711.973338  9173.530539
11  23711.973338  9173.530539
12  23711.973338  9173.530539
13  23711.973338  9173.530539
14  23711.973338  9173.530539
15  23711.973338  9173.530539
16  23711.973338  9173.530539
17  23711.973338  9173.530539
18  23711.973338  9173.530539
19  23711.973338  9173.530539
20  23711.973338  9173.530539
21  23711.973338  9173.530539
22  23711.973338  9173.530539
23  23711.973338  9173.530539
24  23711.973338  9173.530539
25  23711.973338  9173.530539
26  23711.973338  9173.530539
27  23711.973338  9173.530539
28  23711.973338  9173.530539
29  23711.973338  9173.530539
30  23711.973338  9173.530539
31  23711.973338  9173.530539
32  23711.973338  9173.530539
33  23711.973338  9173.530539
34  23711.973338  9173.530539
35  23711.973338  9173.530539
36  23711.973338  9173.530539
37  23711.973338  9173.530539
38  23711.973338  9173.530539
39  23711.973338  9173.530539
40  23711.973338  9173.530539
41  23711.973338  9173.530539
42  23711.973338  9173.530539

In [58]:

print(f'RMSE of all next 42 days: {error_df["RMSE"].sum()} on X_train')
print(f'MAE of all next 42 days: {error_df["MAE"].sum()} on X_train')

RMSE of all next 42 days: 995902.8801970452 on X_train
MAE of all next 42 days: 385288.2826215123 on X_train


### Doing prediction on X_test using beta_all learned previously. Then finding both rmse and mae error and display it

In [60]:
predictions_Test = np.dot(X_test,beta_All.T)
error_test_set= np.zeros(shape=(42,2))
for i in range(42):
    numerator = ((Y_test.iloc[:,i] - predictions_Test[:,i])**2).sum()
    rmse = sqrt((numerator)/predictions_Test.shape[0])
    error_test_set[i][0] = rmse
    
    numerator_mae = (np.absolute(Y_test.iloc[:,i] - predictions_Test[:,i])).sum()
    mae = numerator_mae/predictions_Test.shape[0]
    error_test_set[i][1] = mae
error_test_df = pd.DataFrame({'RMSE': error_test_set[:, 0], 'MAE': error_test_set[:, 1]})
error_test_df.index = error_test_df.index+1
error_test_df

RMSE            MAE
1   2.406197e+06  358694.896241
2   2.405780e+06  360582.625482
3   2.406170e+06  358681.690971
4   2.406130e+06  358653.745970
5   2.406162e+06  358679.358102
6   2.406134e+06  358732.127994
7   2.406121e+06  358646.255943
8   2.406234e+06  358761.907094
9   2.405888e+06  360588.382004
10  2.406457e+06  359367.533128
11  2.406336e+06  359301.630090
12  2.406243e+06  358799.062219
13  2.406331e+06  358712.810515
14  2.406216e+06  358722.977579
15  2.406234e+06  358713.825192
16  2.405733e+06  360570.582004
17  2.406488e+06  358755.948858
18  2.406263e+06  358695.765963
19  2.406342e+06  358627.774368
20  2.406294e+06  358669.626833
21  2.406201e+06  358575.907241
22  2.406321e+06  358812.799773
23  2.405814e+06  360574.573309
24  2.406490e+06  359112.922058
25  2.406338e+06  358834.222557
26  2.406335e+06  358809.438903
27  2.406417e+06  358806.265759
28  2.406386e+06  358673.019134
29  2.406329e+06  358734.665110
30  2.405994e+06  360588.721135
31  2.406441e+06  358727.067761
32  2.406397e+06  358763.226541
33  2.406163e+06  358668.574261
34  2.406535e+06  358738.723357
35  2.406269e+06  358639.919072
36  2.406333e+06  358872.296777
37  2.405969e+06  360585.755917
38  2.406669e+06  359071.452493
39  2.406405e+06  358772.391643
40  2.406438e+06  358688.998509
41  2.406512e+06  358738.774124
42  2.406449e+06  358918.040601

In [62]:
print(f'RMSE of all next 42 days: {error_test_df["RMSE"].sum()} on X_test')
print(f'MAE of all next 42 days: {error_test_df["MAE"].sum()} on X_test')


RMSE of all next 42 days: 101062960.70213382 on X_test
MAE of all next 42 days: 15079666.282587657 on X_test


<hr style="border:2px solid gray"> </hr><hr style="border:2px solid gray"> </hr>

### G.c)

### sorting dataframe by row

In [63]:
df_c = df2[["Store","Sales","Date"]].sort_values(by=['Store','Date'],ascending=True)

### add new coloumn in dataframe which is day name in date coloumn

In [64]:
df_c['Day'] = pd.DatetimeIndex(df_c['Date']).day_name()
df_c.head(n=3)

Store  Sales        Date        Day
941      1      0  2013-01-01    Tuesday
940      1   5530  2013-01-02  Wednesday
939      1   4327  2013-01-03   Thursday

### groupby first on day, then on store to get sale of each store for each day in week

In [65]:
sales_gc = df_c.groupby(['Day','Store']).mean()

In [66]:
# resetting index to convert multi index to single index
rsd = sales_gc.reset_index()  

In [67]:
rsd

Day  Store         Sales
0        Friday      1   4516.414815
1        Friday      2   4462.051852
2        Friday      3   6820.222222
3        Friday      4   9049.503704
4        Friday      5   4606.155556
...         ...    ...           ...
7800  Wednesday   1111   5000.866667
7801  Wednesday   1112   9540.637037
7802  Wednesday   1113   6308.622222
7803  Wednesday   1114  19082.674074
7804  Wednesday   1115   5666.155556

[7805 rows x 3 columns]

### For each store get its store id and day name. Then use  same store_id and day_name to get sales of that day from rsd dataframe (made in previous step) and save these sales in dictionary with id same as storeId. Each item of dictionary is store id and mean of all sales recorded for that day of week in the past

In [83]:
c=1
dict1 = {}
for i in range(1,1116):    
    sds = df_c[df_c["Store"]==i]
#     datess = df_c[df_c["Store"]==i]["Day"]
    datess = df_c[(df_c["Store"]==i) & (df_c["Day"] != "Sunday")]["Day"]
    
    mean_of_sales = []
    for k in range(datess.shape[0]):
        day_cur = datess.iloc[k]
        sale_on_that_day = rsd[(rsd["Store"] == i) & (rsd["Day"] ==day_cur)]
        mean_of_sales.append(sale_on_that_day.values[0][-1])
    
    
    dict1[c] = mean_of_sales
    c= c+1
    print("Done Store: ",i)


Done Store:  1
Done Store:  2
Done Store:  3
Done Store:  4
Done Store:  5
Done Store:  6
Done Store:  7
Done Store:  8
Done Store:  9
Done Store:  10
Done Store:  11
Done Store:  12
Done Store:  13
Done Store:  14
Done Store:  15
Done Store:  16
Done Store:  17
Done Store:  18
Done Store:  19
Done Store:  20
Done Store:  21
Done Store:  22
Done Store:  23
Done Store:  24
Done Store:  25
Done Store:  26
Done Store:  27
Done Store:  28
Done Store:  29
Done Store:  30
Done Store:  31
Done Store:  32
Done Store:  33
Done Store:  34
Done Store:  35
Done Store:  36
Done Store:  37
Done Store:  38
Done Store:  39
Done Store:  40
Done Store:  41
Done Store:  42
Done Store:  43
Done Store:  44
Done Store:  45
Done Store:  46
Done Store:  47
Done Store:  48
Done Store:  49
Done Store:  50
Done Store:  51
Done Store:  52
Done Store:  53
Done Store:  54
Done Store:  55
Done Store:  56
Done Store:  57
Done Store:  58
Done Store:  59
Done Store:  60
Done Store:  61
Done Store:  62
Done Store:  63
D

### converting dictionary to dataframe

In [85]:
df_fin = pd.DataFrame.from_dict(dict1,orient='index')
df_4_fin = df_fin.T
df_4_fin.index = df_4_fin.index + 1
df_5_fin = df_4_fin.T
df_5_fin

1             2             3             4             5    \
1      4650.918519   4454.474074   4094.607407   4516.414815   4942.970149   
2      5329.429630   5705.281481   4673.103704   4462.051852   2873.104478   
3      7588.214815   6973.688889   6371.244444   6820.222222   4434.641791   
4      9340.200000   8721.762963   8535.222222   9049.503704  10113.447761   
5      5071.844444   5019.044444   4431.429630   4606.155556   2079.104478   
...            ...           ...           ...           ...           ...   
1111   5632.162963   5000.866667   4681.303704   5104.651852   3288.723881   
1112  10131.555556   9540.637037   8879.466667   9455.659259   8592.388060   
1113   6512.562963   6308.622222   6570.962963   6293.859259   5893.574627   
1114  19855.244444  19082.674074  18593.325926  19969.533333  21935.500000   
1115   5811.533333   5666.155556   5396.607407   6208.881481   6773.388060   

               6             7             8             9             10   \
1      4946.119403   4650.918519   4454.474074   4094.607407   4516.414815   
2      5790.522388   5329.429630   5705.281481   4673.103704   4462.051852   
3      7965.029851   7588.214815   6973.688889   6371.244444   6820.222222   
4     10365.686567   9340.200000   8721.762963   8535.222222   9049.503704   
5      5834.880597   5071.844444   5019.044444   4431.429630   4606.155556   
...            ...           ...           ...           ...           ...   
1111   6670.328358   5632.162963   5000.866667   4681.303704   5104.651852   
1112  12626.313433  10131.555556   9540.637037   8879.466667   9455.659259   
1113   7006.656716   6512.562963   6308.622222   6570.962963   6293.859259   
1114  20900.171642  19855.244444  19082.674074  18593.325926  19969.533333   
1115   6704.223881   5811.533333   5666.155556   5396.607407   6208.881481   

      ...           799           800           801           802  \
1     ...   4650.918519   4454.474074   4094.607407   4516.414815   
2     ...   5329.429630   5705.281481   4673.103704   4462.051852   
3     ...   7588.214815   6973.688889   6371.244444   6820.222222   
4     ...   9340.200000   8721.762963   8535.222222   9049.503704   
5     ...   5071.844444   5019.044444   4431.429630   4606.155556   
...   ...           ...           ...           ...           ...   
1111  ...   5632.162963   5000.866667   4681.303704   5104.651852   
1112  ...  10131.555556   9540.637037   8879.466667   9455.659259   
1113  ...   6512.562963   6308.622222   6570.962963   6293.859259   
1114  ...  19855.244444  19082.674074  18593.325926  19969.533333   
1115  ...   5811.533333   5666.155556   5396.607407   6208.881481   

               803           804           805           806           807  \
1      4942.970149   4946.119403   4650.918519   4454.474074   4094.607407   
2      2873.104478   5790.522388   5329.429630   5705.281481   4673.103704   
3      4434.641791   7965.029851   7588.214815   6973.688889   6371.244444   
4     10113.447761  10365.686567   9340.200000   8721.762963   8535.222222   
5      2079.104478   5834.880597   5071.844444   5019.044444   4431.429630   
...            ...           ...           ...           ...           ...   
1111   3288.723881   6670.328358   5632.162963   5000.866667   4681.303704   
1112   8592.388060  12626.313433  10131.555556   9540.637037   8879.466667   
1113   5893.574627   7006.656716   6512.562963   6308.622222   6570.962963   
1114  21935.500000  20900.171642  19855.244444  19082.674074  18593.325926   
1115   6773.388060   6704.223881   5811.533333   5666.155556   5396.607407   

               808  
1      4516.414815  
2      4462.051852  
3      6820.222222  
4      9049.503704  
5      4606.155556  
...            ...  
1111   5104.651852  
1112   9455.659259  
1113   6293.859259  
1114  19969.533333  
1115   6208.881481  

[1115 rows x 808 columns]

### Test train split

In [100]:
X_train = df_5_fin.iloc[:1000,:700]
Y_train = df_5_fin.iloc[:1000,700:]

X_test = df_5_fin.iloc[1000:,:700]
Y_test = df_5_fin.iloc[1000:,700:]

X_train.fillna(0,inplace=True)
Y_train.fillna(0,inplace=True)
X_test.fillna(0,inplace=True)
Y_test.fillna(0,inplace=True)

### Prediction on test set

In [99]:
X_transpose = X_train.T
A = np.dot(X_transpose , X_train)

# beta_All = []
beta_All = np.zeros(shape=(108,700))

try:
    for i in range(108):
        b = np.dot(X_transpose , Y_train.iloc[:,i])
        
        inverse_A = inv(A)
        beta_hat = np.dot(inverse_A,b)
        beta_All[i] = beta_hat
except:
    print("Matrix Becomes singular!!! Unable to do inverse")

Matrix Becomes singular!!! Unable to do inverse


#### unable to do prediction on using baseline approach C becuase matric becomes singular while taking inverse!!!

### Comparison of baseline approach with multi regression using normal equation

<p>On X_test: In case of simple multi regression using normal equation: aggeregate of RMSE of all next 42 days: was 105986420.26405494, aggeregate of MAE of all next 42 days: 15641897.644539073</p>

<p>On X_test: In case of baseline approach A : aggeregate of RMSE of all next 42 days: was 113839908.54505162, aggeregate of MAE of all next 42 days: 17071206.73044961</p>

<p>On X_test: In case of baseline approach B : aggeregate of RMSE of all next 42 days: was 101062960.70213382, aggeregate of MAE of all next 42 days: 15079666.282587657</p>

<p>Above results show that in case of base line approach B and C, the aggereate RMSE and MAE is less than RMSE and MAE of simple multi regression. This is becuase the given problem was  time series auto regression problem instead of simple uni-variate regression. In the current problem, when we started using previous data as predictors of next days (baseline approach, the both errors were reduced.</p>

<p></p>


### H.)


<p>Linear regression is not a good choice for this task because this task is classicial example for Auto-regression. In this case, prediction is time dependent. To predict next day sales, model should take into account previous predictions. In auto regression case, model Y is based on previous values of Y. Our example is similar to this where we are predicting sales of next 42 days. Our model should take into considerations the previous predicted sales. </p>
    
<p>In case of linear regression, the prediction is independent of time and it does not take into account previous predictions.In this case, model Y is based on distribution of X</p>

<p>so for this task linear regresssion is a bad choice</p>